### ISSUE TITLE
Allow resources as span attributes

### ISSUE THREAD

**Issue by Oberon00** [open-telemetry/opentelemetry-specification#274]
See https://github.com/open-telemetry/opentelemetry-specification/pull/263#discussion_r329281769.

For some types of spans (e.g. HTTP server) there is potentially a lot of information that can be different from span to span, but there is only a low number of combinations for each process. For example, it would make sense to have `http.app_info` attribute the value of which is a resource that contains information such as an application root, server name, application name, server port (see linked PR).

**Comment by bogdandrutu** [comment]
Please clarify what an "Application" is. Is that a "http.route"?

**Comment by Oberon00** [comment]
See my new comments: https://github.com/open-telemetry/opentelemetry-specification/pull/263#discussion_r329726240

**Comment by Oberon00** [comment]
Some clarifications (from https://github.com/open-telemetry/opentelemetry-specification/pull/263#discussion_r336907328):

> Basically the problem with the app attributes is that in many scenarios (admittedly, probably most) they will have the same values every time they are set. However, I think there are scenarios in which there are multiple apps within the same process, and that's why I disagree with your statement that these attributes "clearly" belong to resources. In these cases we can't use the resources concept which is process-wide, because we have no way to indicate which app the span belongs too.
>
> My goal with #274 would be to have just a single attribute "http.app_info" or similar the value of which is a (reference to a) process-wide app resource. Then you can use the process-wide resources to enumerate all the apps in the process and still know to which app the span belongs. Logically, I think that the app attributes would be perfectly fine as-is, but from a performance-perspective I can see that having them grouped together or transferred as a resource reference can have advantages.
>
> Of course that attribute would not be inherited, just like a string attribute is not inherited.

**Comment by carlosalberto** [comment]
Hey @Oberon00 what's the state of this? can we close it? If not, what remains to be done?

**Comment by Oberon00** [comment]
Well, https://github.com/open-telemetry/opentelemetry-specification/pull/421/ and open-telemetry/oteps#66 could both be helpful when designing a solution, but nothing has been done explicitly to solve this issue and the initial description still applies.

I think of this primarily as a potential solution to #335, which I would still love to be solved.

**Comment by jmacd** [comment]
Related proposal: https://github.com/open-telemetry/oteps/issues/78

**Comment by carlosalberto** [comment]
Moving this to the next milestone.

**Comment by tsloughter** [comment]
This sounds more like having complex values for attributes and not necessarily resources?

I like @jmacd's otep. For a simpler approach I was going to propose being able to attach a Resource to either the context or named tracer.

**Comment by Oberon00** [comment]
@tsloughter Why complex values? This is only about not needing to send the data for each Span but only once per batch (or less).

**Comment by tsloughter** [comment]
@Oberon00 ah, didn't consider batching since it would be on individual spans. Now I get it why it is more than just key/value pairs for the value of an attribute.

I'd like something like this, or jmacd's otep (or just tie the resource to a context so all spans in that context use it). My use is processes in Erlang nodes that are long lived and can represent their own "service". So would want to have spans created in that process have a resource with `serivce.name` specific to the process.

**Comment by svrnm** [comment]
The label removal we applied during triage stopped the `stale` bot from closing this issue, so we close it manually, if this is incorrect please re-open or raise a new issue, cc @mx-psi @jpkrohling
